File: internal/emitter/emitter.go
package emitter

import (
	"fmt"
	"strings"
	"syscall"
	"unicode/utf8"
	"unsafe"

	"hanfe/internal/linux"
	"hanfe/internal/ttybridge"
	"hanfe/internal/util"
)

type FallbackEmitter struct {
	uinputFD     int
	ttyClient    *ttybridge.Client
	ptyFD        int
	closed       bool
	hexKeycodes  [16]int
	inputBuffer  strings.Builder
	directCommit bool
}

const (
	absCnt = 0x3f + 1
)

type inputID struct {
	Bustype uint16
	Vendor  uint16
	Product uint16
	Version uint16
}

type uinputUserDev struct {
	Name         [linux.UinputMaxNameSize]byte
	ID           inputID
	FFEffectsMax int32
	Absmax       [absCnt]int32
	Absmin       [absCnt]int32
	Absfuzz      [absCnt]int32
	Absflat      [absCnt]int32
}

func Open(hexMap map[rune]uint16, ttyClient *ttybridge.Client, ptyPath string, directCommit bool) (*FallbackEmitter, error) {
	emitter := &FallbackEmitter{uinputFD: -1, ttyClient: nil, ptyFD: -1, directCommit: directCommit}
	for i := range emitter.hexKeycodes {
		emitter.hexKeycodes[i] = -1
	}
	for ch, code := range hexMap {
		idx := hexIndex(ch)
		if idx >= 0 && idx < len(emitter.hexKeycodes) {
			emitter.hexKeycodes[idx] = int(code)
		}
	}

	if directCommit {
		emitter.ttyClient = ttyClient
	} else {
		if ttyClient != nil {
			_ = ttyClient.Close()
		}
		fd, err := syscall.Open("/dev/uinput", syscall.O_WRONLY|syscall.O_NONBLOCK|syscall.O_CLOEXEC, 0)
		if err != nil {
			return nil, fmt.Errorf("open /dev/uinput: %w", err)
		}
		emitter.uinputFD = fd

		if err := configureUinput(fd); err != nil {
			syscall.Close(fd)
			return nil, err
		}
	}

	if directCommit && ptyPath != "" {
		ptyFD, err := syscall.Open(ptyPath, syscall.O_WRONLY|syscall.O_CLOEXEC, 0)
		if err != nil {
			emitter.Close()
			return nil, fmt.Errorf("open pty %s: %w", ptyPath, err)
		}
		emitter.ptyFD = ptyFD
	}

	return emitter, nil
}

func configureUinput(fd int) error {
	if err := linux.IoctlSetInt(fd, linux.UISetEvbit, linux.EvSyn); err != nil {
		return fmt.Errorf("UI_SET_EVBIT(EV_SYN): %w", err)
	}
	if err := linux.IoctlSetInt(fd, linux.UISetEvbit, linux.EvKey); err != nil {
		return fmt.Errorf("UI_SET_EVBIT(EV_KEY): %w", err)
	}
	for code := 0; code <= linux.KeyMax; code++ {
		_ = linux.IoctlSetInt(fd, linux.UISetKeybit, code)
	}

	var setup uinputUserDev
	copy(setup.Name[:], []byte("hanfe-fallback"))
	setup.ID.Bustype = linux.BusUSB
	setup.ID.Vendor = 0x1
	setup.ID.Product = 0x1
	setup.ID.Version = 1

	size := unsafe.Sizeof(setup)
	buf := linux.UnsafeSlice((*byte)(unsafe.Pointer(&setup)), int(size))
	if _, err := syscall.Write(fd, buf); err != nil {
		return fmt.Errorf("write uinput setup: %w", err)
	}

	if err := linux.IoctlSetInt(fd, linux.UIDevCreate, 0); err != nil {
		return fmt.Errorf("UI_DEV_CREATE: %w", err)
	}
	return nil
}

func (e *FallbackEmitter) Close() error {
	if e.closed {
		return nil
	}
	e.closed = true
	_ = e.flushBuffer()
	if e.uinputFD >= 0 {
		_ = linux.IoctlSetInt(e.uinputFD, linux.UIDevDestroy, 0)
		syscall.Close(e.uinputFD)
		e.uinputFD = -1
	}
	if e.ttyClient != nil {
		_ = e.ttyClient.Close()
		e.ttyClient = nil
	}
	if e.ptyFD >= 0 {
		syscall.Close(e.ptyFD)
		e.ptyFD = -1
	}
	return nil
}

func (e *FallbackEmitter) ForwardEvent(ev *util.InputEvent) error {
	if e.uinputFD < 0 || ev == nil {
		return nil
	}
	if _, err := syscall.Write(e.uinputFD, ev.Bytes()); err != nil {
		return err
	}
	return e.emitSync()
}

func (e *FallbackEmitter) SendKeyState(code uint16, pressed bool) error {
	value := int32(0)
	if pressed {
		value = 1
	}
	event := util.InputEvent{Type: linux.EvKey, Code: code, Value: value}
	if _, err := syscall.Write(e.uinputFD, event.Bytes()); err != nil {
		return err
	}
	return e.emitSync()
}

func (e *FallbackEmitter) TapKey(code uint16) error {
	if err := e.SendKeyState(code, true); err != nil {
		return err
	}
	return e.SendKeyState(code, false)
}

func (e *FallbackEmitter) emitSync() error {
	if e.uinputFD < 0 {
		return nil
	}
	syn := util.InputEvent{Type: linux.EvSyn, Code: linux.SynReport, Value: 0}
	_, err := syscall.Write(e.uinputFD, syn.Bytes())
	return err
}

func (e *FallbackEmitter) SendBackspace(count int) error {
	if err := e.flushBuffer(); err != nil {
		return err
	}
	for i := 0; i < count; i++ {
		if !e.directCommit {
			if err := e.TapKey(uint16(linux.KeyBackspace)); err != nil {
				return err
			}
		}
		if err := e.mirrorBackspace(); err != nil {
			return err
		}
	}
	return nil
}

func (e *FallbackEmitter) SendText(text string) error {
	if text == "" {
		return nil
	}
	e.inputBuffer.WriteString(text)
	if e.directCommit {
		if !utf8.ValidString(text) {
			e.inputBuffer.Reset()
			return fmt.Errorf("invalid utf-8 sequence")
		}
		return e.flushBuffer()
	}
	remaining := text
	for len(remaining) > 0 {
		r, size := utf8.DecodeRuneInString(remaining)
		if r == utf8.RuneError && size == 1 {
			e.inputBuffer.Reset()
			return fmt.Errorf("invalid utf-8 sequence")
		}
		if err := e.typeUnicode(r); err != nil {
			e.inputBuffer.Reset()
			return err
		}
		remaining = remaining[size:]
	}
	return e.flushBuffer()
}

func (e *FallbackEmitter) ptyWriteBytes(data string) error {
	if e.ptyFD < 0 || data == "" {
		return nil
	}
	buf := []byte(data)
	for len(buf) > 0 {
		n, err := syscall.Write(e.ptyFD, buf)
		if err != nil {
			if err == syscall.EINTR {
				continue
			}
			return err
		}
		buf = buf[n:]
	}
	return nil
}

func (e *FallbackEmitter) ptyPushByte(b byte) error {
	if e.ptyFD < 0 {
		return nil
	}
	buf := []byte{b}
	for len(buf) > 0 {
		n, err := syscall.Write(e.ptyFD, buf)
		if err != nil {
			if err == syscall.EINTR {
				continue
			}
			return err
		}
		buf = buf[n:]
	}
	return nil
}

func (e *FallbackEmitter) mirrorWrite(data string) error {
	if err := e.ttyClient.WriteString(data); err != nil {
		return err
	}
	if err := e.ptyWriteBytes(data); err != nil {
		return err
	}
	return nil
}

func (e *FallbackEmitter) mirrorBackspace() error {
	if err := e.ttyClient.SendBackspace(1); err != nil {
		return err
	}
	if err := e.ptyPushByte('\b'); err != nil {
		return err
	}
	return nil
}

func (e *FallbackEmitter) flushBuffer() error {
	if e.inputBuffer.Len() == 0 {
		return nil
	}
	data := e.inputBuffer.String()
	e.inputBuffer.Reset()
	if data == "" {
		return nil
	}
	return e.mirrorWrite(data)
}

func (e *FallbackEmitter) typeUnicode(r rune) error {
	if e.uinputFD < 0 {
		return nil
	}
	if err := e.SendKeyState(uint16(linux.KeyLeftCtrl), true); err != nil {
		return err
	}
	if err := e.SendKeyState(uint16(linux.KeyLeftShift), true); err != nil {
		return err
	}
	if err := e.TapKey(uint16(linux.KeyU)); err != nil {
		return err
	}
	if err := e.SendKeyState(uint16(linux.KeyLeftShift), false); err != nil {
		return err
	}
	if err := e.SendKeyState(uint16(linux.KeyLeftCtrl), false); err != nil {
		return err
	}

	hexDigits := fmt.Sprintf("%x", r)
	for _, ch := range hexDigits {
		idx := hexIndex(ch)
		if idx < 0 || idx >= len(e.hexKeycodes) {
			continue
		}
		key := e.hexKeycodes[idx]
		if key < 0 {
			continue
		}
		if err := e.TapKey(uint16(key)); err != nil {
			return err
		}
	}

	if err := e.SendKeyState(uint16(linux.KeyLeftCtrl), true); err != nil {
		return err
	}
	if err := e.SendKeyState(uint16(linux.KeyLeftShift), true); err != nil {
		return err
	}
	if err := e.TapKey(uint16(linux.KeyEnter)); err != nil {
		return err
	}
	if err := e.SendKeyState(uint16(linux.KeyLeftShift), false); err != nil {
		return err
	}
	if err := e.SendKeyState(uint16(linux.KeyLeftCtrl), false); err != nil {
		return err
	}
	return nil
}

func hexIndex(ch rune) int {
	switch {
	case ch >= '0' && ch <= '9':
		return int(ch - '0')
	case ch >= 'a' && ch <= 'f':
		return 10 + int(ch-'a')
	case ch >= 'A' && ch <= 'F':
		return 10 + int(ch-'A')
	default:
		return -1
	}
}

func (e *FallbackEmitter) SupportsPreedit() bool {
	return e.directCommit
}

File: src/emitter.cpp
#include "hanfe/emitter.hpp"

#include <fcntl.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include <cerrno>
#include <cstdio>
#include <cstring>
#include <cstdint>
#include <iomanip>
#include <sstream>
#include <stdexcept>

#include "hanfe/util.hpp"

namespace hanfe {
namespace {

constexpr int kMaxKeyCode = KEY_MAX;

void set_keybits(int fd) {
    if (ioctl(fd, UI_SET_EVBIT, EV_SYN) < 0 || ioctl(fd, UI_SET_EVBIT, EV_KEY) < 0) {
        throw std::runtime_error("Failed to configure uinput event bits: " +
                                 std::string(strerror(errno)));
    }
    for (int code = 0; code <= kMaxKeyCode; ++code) {
        ioctl(fd, UI_SET_KEYBIT, code);
    }
}

}  // namespace

FallbackEmitter::FallbackEmitter(const std::unordered_map<char, int>& hex_keys,
                                 std::optional<std::string> tty_path)
    : hex_keys_(hex_keys) {
    uinput_fd_ = ::open("/dev/uinput", O_WRONLY | O_NONBLOCK | O_CLOEXEC);
    if (uinput_fd_ < 0) {
        throw std::runtime_error("Failed to open /dev/uinput: " + std::string(strerror(errno)));
    }

    set_keybits(uinput_fd_);

    struct uinput_user_dev setup;
    std::memset(&setup, 0, sizeof(setup));
    std::snprintf(setup.name, UINPUT_MAX_NAME_SIZE, "hanfe-fallback");
    setup.id.bustype = BUS_USB;
    setup.id.vendor = 0x1;
    setup.id.product = 0x1;
    setup.id.version = 1;

    if (write(uinput_fd_, &setup, sizeof(setup)) < 0) {
        throw std::runtime_error("Failed to configure uinput device: " +
                                 std::string(strerror(errno)));
    }
    if (ioctl(uinput_fd_, UI_DEV_CREATE) < 0) {
        throw std::runtime_error("Failed to create uinput device: " +
                                 std::string(strerror(errno)));
    }

    if (tty_path) {
        int fd = ::open(tty_path->c_str(), O_WRONLY | O_NOCTTY | O_CLOEXEC);
        if (fd < 0) {
            throw std::runtime_error("Failed to open TTY '" + *tty_path + "': " +
                                     std::string(strerror(errno)));
        }
        tty_fd_ = fd;
    }
}

FallbackEmitter::~FallbackEmitter() { close(); }

void FallbackEmitter::close() {
    if (closed_) {
        return;
    }
    closed_ = true;
    if (uinput_fd_ >= 0) {
        ioctl(uinput_fd_, UI_DEV_DESTROY);
        ::close(uinput_fd_);
        uinput_fd_ = -1;
    }
    if (tty_fd_) {
        ::close(*tty_fd_);
        tty_fd_.reset();
    }
}

void FallbackEmitter::emit(unsigned int type, unsigned int code, int value) {
    if (uinput_fd_ < 0) {
        return;
    }
    input_event ev{};
    ev.type = type;
    ev.code = static_cast<unsigned short>(code);
    ev.value = value;
    if (write(uinput_fd_, &ev, sizeof(ev)) < 0) {
        throw std::runtime_error("Failed to write to uinput: " + std::string(strerror(errno)));
    }
    input_event syn{};
    syn.type = EV_SYN;
    syn.code = SYN_REPORT;
    syn.value = 0;
    if (write(uinput_fd_, &syn, sizeof(syn)) < 0) {
        throw std::runtime_error("Failed to sync uinput: " + std::string(strerror(errno)));
    }
}

void FallbackEmitter::forward_event(const input_event& event) {
    emit(event.type, event.code, event.value);
}

void FallbackEmitter::send_key_state(int keycode, bool pressed) {
    emit(EV_KEY, keycode, pressed ? 1 : 0);
}

void FallbackEmitter::tap_key(int keycode) {
    send_key_state(keycode, true);
    send_key_state(keycode, false);
}

void FallbackEmitter::send_backspace(int count) {
    for (int i = 0; i < count; ++i) {
        tap_key(KEY_BACKSPACE);
        write_tty("\b");
    }
}

void FallbackEmitter::send_text(const std::string& text) {
    if (text.empty()) {
        return;
    }
    std::u32string codepoints = utf8_to_u32(text);
    for (char32_t cp : codepoints) {
        type_unicode(cp);
        write_tty(utf8_from_char32(cp));
    }
}

// Helper: convert codepoint -> UTF-8
static std::string codepoint_to_utf8(char32_t cp) {
    std::string out;
    if (cp <= 0x7F) {
        out.push_back(static_cast<char>(cp));
    } else if (cp <= 0x7FF) {
        out.push_back(static_cast<char>(0xC0 | ((cp >> 6) & 0x1F)));
        out.push_back(static_cast<char>(0x80 | (cp & 0x3F)));
    } else if (cp <= 0xFFFF) {
        out.push_back(static_cast<char>(0xE0 | ((cp >> 12) & 0x0F)));
        out.push_back(static_cast<char>(0x80 | ((cp >> 6) & 0x3F)));
        out.push_back(static_cast<char>(0x80 | (cp & 0x3F)));
    } else {
        out.push_back(static_cast<char>(0xF0 | ((cp >> 18) & 0x07)));
        out.push_back(static_cast<char>(0x80 | ((cp >> 12) & 0x3F)));
        out.push_back(static_cast<char>(0x80 | ((cp >> 6) & 0x3F)));
        out.push_back(static_cast<char>(0x80 | (cp & 0x3F)));
    }
    return out;
}

void FallbackEmitter::type_unicode(char32_t codepoint) {
    // keep original modifier+U sequence so GUI fallback stays intact
    send_key_state(KEY_LEFTCTRL, true);
    send_key_state(KEY_LEFTSHIFT, true);
    tap_key(KEY_U);
    send_key_state(KEY_LEFTSHIFT, false);
    send_key_state(KEY_LEFTCTRL, false);

    // write the real UTF-8 to TTY for visibility/logging
    write_tty(codepoint_to_utf8(codepoint));

    // *** If a TTY is open, DO NOT type hex digits (they would duplicate / conflict) ***
    if (tty_fd_) {
        return;
    }

    // Otherwise (no tty), fall back to typing hex digits as before
    std::ostringstream oss;
    oss << std::hex << std::nouppercase << static_cast<uint32_t>(codepoint);
    std::string hex = oss.str();

    for (char ch : hex) {
        auto it = hex_keys_.find(ch);
        if (it == hex_keys_.end()) continue;
        tap_key(it->second);
    }
}

void FallbackEmitter::write_tty(const std::string& text) {
    if (!tty_fd_ || text.empty()) {
        return;
    }
    ssize_t ignored = ::write(*tty_fd_, text.data(), text.size());
    (void)ignored;
}

}  // namespace hanfe
